File: solutions/go/spiral-matrix/1/spiral_matrix.go
// Package spiralmatrix ...
package spiralmatrix

// SpiralMatrix ...
func SpiralMatrix(size int) [][]int {
	m := make([][]int, size)
	for i := range m {
		m[i] = make([]int, size)
	}
	top, bottom := 0, size-1
	left, right := 0, size-1
	for val := 1; val <= size*size; {
		for i := left; i <= right; i++ {
			m[top][i] = val
			val++
		}
		top++

		for i := top; i <= bottom; i++ {
			m[i][right] = val
			val++
		}
		right--

		for i := right; i >= left; i-- {
			m[bottom][i] = val
			val++
		}
		bottom--

		for i := bottom; i >= top; i-- {
			m[i][left] = val
			val++
		}
		left++
	}
	return m
}

File: solutions/lua/spiral-matrix/1/spiral-matrix.lua
return function(size)
  local m = {}
  for i = 1, size do m[i] = {} end

  local top, bottom = 1, size
  local left, right = 1, size
  local val = 1
  while val <= size * size do
    for i = left, right do
      m[top][i] = val
      val = val + 1
    end
    top = top + 1

    for i = top, bottom do
      m[i][right] = val
      val = val + 1
    end
    right = right - 1

    for i = right, left, -1 do
      m[bottom][i] = val
      val = val + 1
    end
    bottom = bottom - 1

    for i = bottom, top, -1 do
      m[i][left] = val
      val = val + 1
    end
    left = left + 1
  end
  return m
end
